mailet/crypto/src/main/java/org/apache/james/transport/KeyStoreHolderFactory.java

```diff
@@ -0,0 +1,137 @@
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.transport;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.security.GeneralSecurityException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
+import java.security.cert.CertificateException;
+
+import jakarta.mail.MessagingException;
+
+import org.apache.commons.io.input.UnsynchronizedBufferedInputStream;
+import org.apache.mailet.MailetConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class KeyStoreHolderFactory {
+    private static final Logger LOGGER = LoggerFactory.getLogger(KeyStoreHolderFactory.class);
+    private static final KeyStoreHolderFactory INSTANCE = new KeyStoreHolderFactory();
+    private static final String KEYSTORE_FILE_TYPE = "keystore";
+    private static final String PEM_FILE_TYPE = "pem";
+
+    private KeyStoreHolderFactory() {
+    }
+
+    public static KeyStoreHolderFactory getInstance() {
+        return INSTANCE;
+    }
+
+    public KeyStoreHolder createKeyStoreHolder(MailetConfig config) throws MessagingException {
+        try {
+            initJCE();
+            String fileType = config.getInitParameter("fileType");
+            if (fileType == null) {
+                fileType = KEYSTORE_FILE_TYPE;
+            }
+            switch (fileType) {
+                case KEYSTORE_FILE_TYPE:
+                    return createFromKeyStoreFile(config);
+                case PEM_FILE_TYPE:
+                    return createFromPemFile(config);
+                default:
+                    throw new RuntimeException("Unsupported file type " + fileType);
+            }
+        } catch (Exception e) {
+            throw new MessagingException("Error loading the trusted certificate store", e);
```

Comment:
Why not runtime like above?

Reply:
In the original code, everything is in this try catch block:

```
try {
    ........
} catch (Exception e) {
    throw new MessagingException("Error loading the trusted certificate store", e);
}
```
As I understand, somewhere on top use MessagingException for a specific purpose.

Reply:
just an usual way to handle default case of switch